File: app/src/main/java/com/example/mobileappdevelopment/Activities/ChooseHuntActivity.java
package com.example.mobileappdevelopment.Activities;

import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.SearchView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mobileappdevelopment.Adapters.RecycleViewAdapter;
import com.example.mobileappdevelopment.DataUtils.Cache;
import com.example.mobileappdevelopment.Model.Hunt;
import com.example.mobileappdevelopment.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ChooseHuntActivity extends AppCompatActivity {

    private List<String> titles = new ArrayList<>();
    private List<String> authors = new ArrayList<>();
    private List<Hunt> hunts = new ArrayList<>();

    private SearchView searchView;
    private ProgressBar progressBar;

    private FirebaseFirestore fb;

    @Override
    protected void onCreate(Bundle savedInstances) {
        super.onCreate(savedInstances);
        setContentView(R.layout.activity_choose_hunt);


        searchView = findViewById(R.id.search_view);
        progressBar = findViewById(R.id.progress_choose_hunt);

        // connect to database
//        fb = FirebaseFirestore.getInstance();


        // fill the list with the right data
        populateList();

//        display();

        // display the list in a recycle layout
        setupQuery();


        RecyclerView recyclerView = findViewById(R.id.rvHunts);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        RecyclerView.Adapter adapter = new RecycleViewAdapter(this, titles, authors);
        recyclerView.setAdapter(adapter);
    }


    public void setupQuery() {
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                showSearched(s);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                showSearched(s);
                return false;
            }

        });
    }

    public void populateList() {

        if (Cache.allHunts == null || Cache.allHunts.isEmpty() || Cache.listUpdated) {
            Cache.query.addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                @Override
                public void onComplete(@NonNull Task<QuerySnapshot> task) {
                    for (QueryDocumentSnapshot query : Objects.requireNonNull(task.getResult())) {
                        String huntStringFromDb = query.getString("HuntFile");
                        Gson gson = new Gson();
                        Hunt hunt = gson.fromJson(huntStringFromDb, Hunt.class);
                        if (!hunt.isPrivate()){
                            Cache.allHuntTitles.add(hunt.getTitle());
                            Cache.allHuntAuthors.add(hunt.getAuthor());
                            Cache.allHunts.add(hunt);
                        } else if (hunt.isPrivate() && hunt.getAuthorID().equals(Cache.account.getId())) {
                            Cache.allHuntTitles.add(hunt.getTitle());
                            Cache.allHuntAuthors.add(hunt.getAuthor());
                            Cache.allHunts.add(hunt);
                        }
                    }
                    progressBar.setVisibility(View.GONE);
                    display();
                }
            });
        } else {
            progressBar.setVisibility(View.GONE);
        }
        hunts = Cache.allHunts;
        titles = Cache.allHuntTitles;
        authors = Cache.allHuntAuthors;
        display();

    }

    public void display() {
        // set up the RecyclerView
        RecyclerView recyclerView = findViewById(R.id.rvHunts);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        RecyclerView.Adapter adapter = new RecycleViewAdapter(this, titles, authors);
        recyclerView.setAdapter(adapter);


//        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
//                1);
//        recyclerView.addItemDecoration(dividerItemDecoration);
    }

    public void showSearched(String s) {
        List<Hunt> dynamic_list = new ArrayList<>();

        for (Hunt hunt : Cache.allHunts) {
            if (hunt.getTitle().toUpperCase().contains(s.toUpperCase()) || hunt.getAuthor().toUpperCase().contains(s.toUpperCase())) {
                dynamic_list.add(hunt);
            }
        }

        titles.clear();
        authors.clear();
        for (Hunt hunt : dynamic_list) {
            titles.add(hunt.getTitle());
            authors.add(hunt.getAuthor());
        }

        display();
//        RecyclerView recyclerView = findViewById(R.id.rvHunts);
//        RecycleViewAdapter adapter = new RecycleViewAdapter(this, titles, authors);
//        recyclerView.setAdapter(adapter);
//        recyclerView.setLayoutManager(new LinearLayoutManager(this));
    }
}

File: app/src/main/java/com/example/mobileappdevelopment/activities/ChooseHuntActivity.kt
package com.example.mobileappdevelopment.activities

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.mobileappdevelopment.adapters.RecycleViewAdapter
import com.example.mobileappdevelopment.utils.Cache
import com.example.mobileappdevelopment.model.Hunt
import com.example.mobileappdevelopment.R
import com.google.gson.Gson
import kotlinx.android.synthetic.main.activity_choose_hunt.*
import java.util.*

class ChooseHuntActivity : AppCompatActivity() {
    private var titles: MutableList<String> = ArrayList()
    private var authors: MutableList<String> = ArrayList()
    private var hunts: List<Hunt> = ArrayList()
    private var searchView: SearchView? = null
    private var progressBar: ProgressBar? = null

    override fun onCreate(savedInstances: Bundle?) {
        super.onCreate(savedInstances)
        setContentView(R.layout.activity_choose_hunt)
        searchView = search_view
        progressBar = progress_choose_hunt

        populateList()

        setupQuery()

        display()

    }

    private fun setupQuery() {
        searchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                showSearched(s)
                return false
            }

            override fun onQueryTextChange(s: String): Boolean {
                showSearched(s)
                return false
            }
        })
    }

    private fun populateList() {
        if (Cache.allHunts.isEmpty() || Cache.listUpdated) {
            Cache.allHunts.clear()
            Cache.allHuntAuthors.clear()
            Cache.allHuntTitles.clear()
            Cache.query?.addOnCompleteListener { task ->
                for (query in task.result!!) {
                    val huntStringFromDb = query.getString("HuntFile")
                    val gson = Gson()
                    val hunt = gson.fromJson(huntStringFromDb, Hunt::class.java)
                    if (!hunt.isPrivate) {
                        hunt.title?.let { Cache.allHuntTitles.add(it) }
                        hunt.author?.let { Cache.allHuntAuthors.add(it) }
                        Cache.allHunts.add(hunt)
                    } else if (hunt.isPrivate && hunt.authorID == Cache.account?.id) {
                        hunt.title?.let { Cache.allHuntTitles.add(it) }
                        hunt.author?.let { Cache.allHuntAuthors.add(it) }
                        Cache.allHunts.add(hunt)
                    }
                }
                progressBar!!.visibility = View.GONE
                display()
            }
        } else {
            progressBar!!.visibility = View.GONE
        }
        hunts = Cache.allHunts
        titles = Cache.allHuntTitles
        authors = Cache.allHuntAuthors
        display()
    }

    // set up the RecyclerView
    private fun display() {
        val recyclerView = findViewById<RecyclerView>(R.id.rvHunts)
        recyclerView.layoutManager = LinearLayoutManager(this)
        val adapter: RecyclerView.Adapter<*> = RecycleViewAdapter(this, titles, authors)
        recyclerView.adapter = adapter
    }

    @SuppressLint("DefaultLocale")
    fun showSearched(s: String) {
        val dynamicList: MutableList<Hunt> = ArrayList()
        for (hunt in Cache.allHunts) {
            if (hunt.title!!.toUpperCase().contains(s.toUpperCase()) || hunt.author!!.toUpperCase().contains(s.toUpperCase())) {
                dynamicList.add(hunt)
            }
        }
        titles.clear()
        authors.clear()
        for (hunt in dynamicList) {
            hunt.title?.let { titles.add(it) }
            hunt.author?.let { authors.add(it) }
        }
        display()

    }
}